design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,287 @@
+arguments and results of an operation.
+ 
+Example on how would we change DTOs for operations regarding customer creation/update:
+ 
+First we would have DTO that describes a customer in the way it's convenient for client code when updating a customer:
+```php
+interface CustomerUpdateInterface
+{
+    //No getId() since storefront clients must not be able to control their customer's ID
+    ...
+    
+    //For setting new password
+    public function getPassword(): ?string;
+    
+    public function getAddresses(): CustomerAddress[];
+}
+```
+ 
+Then we would have DTO for customer's addresses:
+```php
+interface CustomerAddressUpdateInterface
+{
+    public function getId(): ?string;
+    ....
+    
+    //No getCustomerId() because we don't give control over that
+    ....
+    
+    public function getPostcode(): string
+}
+```
+
+Since we don't want original password in customer responses we would not return it as read operations result:
+```php
+interface CustomerReadInterface
+{
+    //Now we have getId() since it's OK for storefront clients to know their customer's ID.
+    public function getId(): string;
+    
+    ....
+    //No getPassword()
+    ....
+}
+```
+ 
+By having these operation-specific DTOs we would only have the data we need as arguments and responses.
+ 
+###### How to introduce it to Magento?
+There are couple of ways we can introduce this concept to Magento:
+* use it for new storefront API
+* introduce new APIs to a module (for in stance _Customer_) in order to provide an example for 3rd-party developers
+* create a devdocs page/update technical guidelines to inform community
+ 
+ 
+#### Introduce reusable user input processors
+Sometimes before passing arguments to a service contracts we would have to perform user input processing that cannot
+be covered with declarative validation or existing ACL policies regarding of our current area (HTML/REST/SOAP/GraphQL).
+ 
+Example is checking whether is a customer allowed to use a saved shipping address for an order by checking whether the
+address was created by the same customer - we'd have to do it programmatically since validation rules wouldn't work
+because of not having access to the full context and our ACL system is not fit to process those cases. HTML controllers
+and GraphQL resolvers would be able to include such logic but it would not be reusable by other areas. This validation
+could be performed by the service contract responsible for posting order but service contracts are not supposed to know
+about authenticated users.
+ 
+The solution to these problems would be to introduce reusable area-agnostic user input processors that would perform
+such actions. These would be arbitrary classes with no common interface but they would be reusable across different areas
+and reside inside a special namespace `Presentation` to inform developers that this classes belong to presentation
```

Comment:
I propose `InputValidator` or `InputValidation` namespace for validators. It's more precise.
Processors are area-specific, so I don't see a reason to group them under one namespace. In the example below, the processor is specifically for `WebApi`.

Reply:
@buskamuza don't want to limit these classes containing presentation area logic to be seen as validators only. Some of them may contain logic related to extracting data from requests, reading headers for instance. Right now such classes just reside inside Model namespace but they are not part of service layer. What about Model\ContextAware? For classes like session managers, classes for generating URLs for users, input validators, access validators etc?